display_list/display_list_benchmarks.cc

```diff
@@ -335,6 +335,86 @@ void BM_DrawRRect(benchmark::State& state,
   canvas_provider->Snapshot(filename);
 }
 
+// Draws a series of "DR" rects of the requested width across
+// the canvas and repeats until `kRRectsToDraw` rects have been drawn.
+//
+// A "DR" rect is a shape consisting of the difference between two
+// rounded rects.
+//
+// Half the drawn DR rects will not have an integral offset.
+void BM_DrawDRRect(benchmark::State& state,
+                   BackendType backend_type,
+                   unsigned attributes,
+                   SkRRect::Type type) {
+  auto canvas_provider = CreateCanvasProvider(backend_type);
+  DisplayListBuilder builder;
+  builder.setAttributesFromPaint(GetPaintForRun(attributes),
+                                 DisplayListOpFlags::kDrawDRRectFlags);
+  AnnotateAttributes(attributes, state, DisplayListOpFlags::kDrawDRRectFlags);
+
+  size_t length = state.range(0);
+  size_t canvas_size = length * 2;
+  canvas_provider->InitializeSurface(canvas_size, canvas_size);
+  auto canvas = canvas_provider->GetSurface()->getCanvas();
+
+  SkVector radii[4];
+  switch (type) {
+    case SkRRect::Type::kSimple_Type:
+      radii[0] = SkVector::Make(5.0f, 5.0f);
+      radii[1] = SkVector::Make(5.0f, 5.0f);
+      radii[2] = SkVector::Make(5.0f, 5.0f);
+      radii[3] = SkVector::Make(5.0f, 5.0f);
+      break;
+    case SkRRect::Type::kNinePatch_Type:
+      radii[0] = SkVector::Make(5.0f, 2.0f);
+      radii[1] = SkVector::Make(3.0f, 2.0f);
+      radii[2] = SkVector::Make(3.0f, 4.0f);
+      radii[3] = SkVector::Make(5.0f, 4.0f);
+      break;
+    case SkRRect::Type::kComplex_Type:
+      radii[0] = SkVector::Make(5.0f, 4.0f);
+      radii[1] = SkVector::Make(4.0f, 5.0f);
+      radii[2] = SkVector::Make(3.0f, 6.0f);
+      radii[3] = SkVector::Make(2.0f, 7.0f);
+      break;
+    default:
+      break;
+  }
+
+  const SkScalar offset = 0.5f;
+  const SkScalar multiplier = length / 16.0f;
+  SkRRect rrect, rrect_2;
+
+  SkVector set_radii[4];
+  for (size_t i = 0; i < 4; i++) {
+    set_radii[i] = radii[i] * multiplier;
+  }
+  rrect.setRectRadii(SkRect::MakeLTRB(0, 0, length, length), set_radii);
+
+  for (size_t i = 0; i < kRRectsToDraw; i++) {
+    rrect.inset(0.1f * length, 0.1f * length, &rrect_2);
```

Comment:
The dx and dy are also added(subtracted) to/from the radii. I think the smallest radius you have is length/8? If so then you might just barely squeak by on not creating a degenerate RRect. Do I read that right?

Reply:
The smallest radius is 2.0f in the nine patch rrect. In the smallest rect we request (16px length) that makes the multiplier 1.0 so the smallest absolute radius is 2.0f.

That makes our rrect degenerate but only in one corner, so we won't be just hitting some optimisation to call drawRect instead. I think it's fine as is, as I can imagine a usecase where one corner is squared off is plausible.

Reply:
If we test squashed corners, then they should be tested in their own benchmarks, not mixed into the results of another.

Reply:
Changed the radii to larger numbers.